### ISSUE TITLE
 [FR] Computing the score function of a marginal density

### ISSUE THREAD

**Issue by L-z-Chen** [pyro-ppl/pyro#2947]
### Feature Description
A module for computing the score function of a marginal density (both discrete and continuous probability distributions). 

The score function is defined as the gradient of log density: `s_p(x) := grad log p(x)` It is known that it can be computed with an analogue of the Fisher's identity  `s_p(x) = E_{Z|x} [grad log p(x,Z)]` (see, e.g., Eq. (2) of <https://arxiv.org/abs/1907.00586>).
A discrete analogue of the Hyvarinen score has also been considered (http://proceedings.mlr.press/v80/yang18c.html). 

Score functions can be used for model evaluation. In particular, [the kernel Stein discrepancy](https://arxiv.org/abs/1602.03253) (KSD) can be implemented with the score function, and it could serve as an alternative model diagnostic to existing tools (e.g., ELBO/WAIC); model evaluation for latent variable models based on KSD is considered here (https://arxiv.org/abs/1907.00586).  There are other applications in [entropy maximization](https://arxiv.org/abs/1910.04302) or [model fitting with score matching](https://arxiv.org/abs/1910.04302).

### Code Snippet
I have implemented a prototype of the above feature below; this is a snippet to compute the score of a continuous density at a single observation point `obs`. 
If this looks a reasonable point to start, I would like to send a PR. 

```python
def continuous_score(model, obs, latent_variables):
    """Calcalate continuous score
    Args:
        model (func): defined by pyro
        obs (torch.tensor): observation value
        latent_variables (dict): latent variables generated by mcmc
    Returns:
        torch.tensor: score
    """
    name_samples = list(latent_variables)
    num_samples = latent_variables[name_samples[0]].shape[0]
    mean_score = []
    for k in range(num_samples):
        for name in name_samples:
            dict_samples = dict([(name, latent_variables[name][k])])
        dict_samples['obs'] = obs
        mean_score.append(conditioned_c_score(model, dict_samples))
    mean_score = torch.stack(mean_score, dim = -1)
    score = torch.mean(mean_score, dim = -1)
    return score


def make_log_joint(model, sample_data, *args, **kwargs):
    '''
    Evaluate score of the joint at obs and latents
    '''
    conditioned_model = poutine.condition(model, data = sample_data)
    trace = poutine.trace(conditioned_model).get_trace(*args, **kwargs)
    return trace.log_prob_sum()

def conditioned_c_score(model, dict_samples):
    '''
    Calculate conditional continuous score
    '''
    dict_samples['obs'] = dict_samples['obs'].clone().detach().requires_grad_(True)
    log_joint = make_log_joint(model, dict_samples)
    d_log = torch.autograd.grad(outputs = log_joint, inputs = dict_samples['obs'],
                            grad_outputs = torch.ones_like(log_joint), allow_unused = False)[0]
    return d_log
 ```
